src/native/minipal/utf8.c

```diff
@@ -0,0 +1,2161 @@
+        if (!self->treatAsLE) ch = (int)(((unsigned int)ch) >> 16);
+        else
+#endif
+        ch = (CHAR16_T)ch;
+
+        pSrc++;
+
+        if (ch <= 0x7F)
+        {
+            continue;
+        }
+
+        LongCode:
+            // use separate helper variables for slow and fast loop so that the jit optimizations
+            // won't get confused about the variable lifetimes
+            if (ch > 0x7FF)
+            {
+                if (InRange(ch, HIGH_SURROGATE_START, LOW_SURROGATE_END))
+                {
+                    // 4 byte encoding - high surrogate + low surrogate
+
+                    int chd = *pSrc;
+                    if (
+                        ch > HIGH_SURROGATE_END ||
+                        !InRange(chd, LOW_SURROGATE_START, LOW_SURROGATE_END))
+                    {
+                        // Back up and drop out to slow loop to figure out error
+                        pSrc--;
+                        break;
+                    }
+                    pSrc++;
+
+                    // byteCount - this byte is compensated by the second surrogate character
+                }
+                byteCount++;
+            }
+            byteCount++;
+
+            // byteCount - the last byte is already included
+        }
+
+        // no pending char at this point
+        ch = 0;
+    }
+
+#if WIN64
+    // check for overflow
+    assert(byteCount >= 0);
+#endif
+    assert(!fallbackUsed || self->buffer.encoder.fallbackCount < 0);
+
+    return byteCount;
+}
+
+static size_t utf16_to_utf8_preallocated(
+    const CHAR16_T* source,
+    size_t sourceLength,
+    char** destination,
+    size_t destinationLength,
+    unsigned int flags)
+{
+    size_t ret;
+    errno = 0;
+
+    if ((int)sourceLength < 0)
+        sourceLength = minipal_wcslen(source) + 1;
+
+    UTF8Encoding enc =
+    {
+        // repeat replacement char (0xFFFD) twice for a surrogate pair
+        .buffer = { .encoder = { .fallbackCount = -1, .fallbackIndex = -1, .strDefault = { 0xFFFD, 0xFFFD, 0 }, .strDefaultLength = 2 } },
+        .useFallback = true,
+#if BIGENDIAN
+        .treatAsLE = (flags & MINIPAL_TREAT_AS_LITTLE_ENDIAN)
+#endif
+    };
+
+#if !BIGENDIAN
+    (void)flags; // unused
+#endif
+
+    ret = GetByteCount(&enc, (CHAR16_T*)source, sourceLength);
+    if (destinationLength)
+    {
+        if ((size_t)ret > destinationLength)
+        {
+            errno = MINIPAL_ERROR_INSUFFICIENT_BUFFER;
+            ret = 0;
+        }
+
+        GetBytes(&enc, (CHAR16_T*)source, sourceLength, (unsigned char*)*destination, ret);
+        if (errno) ret = 0;
+    }
+
+    return ret;
+}
+
+size_t minipal_utf8_to_utf16_preallocated(
+    const char* source,
+    size_t sourceLength,
+    CHAR16_T** destination,
+    size_t destinationLength,
+    unsigned int flags)
+{
+    size_t ret;
+    errno = 0;
+
+    if ((int)sourceLength < 0)
+        sourceLength = (int)strlen(source) + 1;
+
```

Comment:
This won't produce expected results when somebody manages to pass in >2GB buffer. These special check for -1 should be done by the caller in the PAL.

Reply:
We should have separate methods for computing the length and performing the conversion itself.

Reply:
This is existing code. All callers need this logic, so I moved it from the caller to the common place.

Reply:
But then it does not make sense for the arguments to be `size_t`. I would move this logic to the callers and have a clean methods that operate on size_t in the minipal.